File: __tests__/bs_ZInt_test.ml
open Jest
open Expect
open Z

let () = 

describe "Zint" (fun () ->
  test "rem" (fun () ->
    expect @@ ZInt.(rem (of_int 7) (of_int 3)) |> toEqual (ZInt.of_int 1));

  test "rem" (fun () ->
    expect @@ ZInt.(rem (of_int 7) (of_int (-3))) |> toEqual (ZInt.of_int 1));

  test "rem" (fun () ->
    expect @@ ZInt.(rem (of_int (-7)) (of_int 3)) |> toEqual (ZInt.of_int (-1)));

  test "rem" (fun () ->
    expect @@ ZInt.(rem (of_int (-7)) (of_int (-3))) |> toEqual (ZInt.of_int (-1)));


  test "ediv_rem 7 3" (fun () ->
    expect @@ ZInt.(ediv_rem (of_int 7) (of_int 3)) |> toEqual (ZInt.(of_int 2, of_int 1)));

  test "ediv_rem 7 (-3)" (fun () ->
    expect @@ ZInt.(ediv_rem (of_int 7) (of_int (-3))) |> toEqual (ZInt.(of_int (-2), of_int 1)));

  test "ediv_rem (-7) 3" (fun () ->
    expect @@ ZInt.(ediv_rem (of_int (-7)) (of_int 3)) |> toEqual (ZInt.(of_int (-3), of_int 2)));

  test "ediv_rem (-7) (-3)" (fun () ->
    expect @@ ZInt.(ediv_rem (of_int (-7)) (of_int (-3))) |> toEqual (ZInt.(of_int (-3), of_int 2)));

  test "ediv_rem (-8) 3" (fun () ->
    expect @@ ZInt.(ediv_rem (of_int (-8)) (of_int 3)) |> toEqual (ZInt.(of_int (-3), of_int 1)));

  test "ediv_rem (-8) (-3)" (fun () ->
    expect @@ ZInt.(ediv_rem (of_int (-8)) (of_int (-3))) |> toEqual (ZInt.(of_int (-3), of_int 1)));

  test "ediv_rem 9 3" (fun () ->
    expect @@ ZInt.(ediv_rem (of_int 9) (of_int 3)) |> toEqual (ZInt.(of_int 3, of_int 0 )));

  test "ediv_rem 9 (-3)" (fun () ->
    expect @@ ZInt.(ediv_rem (of_int 9) (of_int (-3))) |> toEqual (ZInt.(of_int (-3), zero )));

  test "ediv_rem (-9) 3" (fun () ->
    expect @@ ZInt.(ediv_rem (of_int (-9)) (of_int 3)) |> toEqual (ZInt.(of_int (-3), of_int 0 )));

  test "ediv_rem (-9) (-3)" (fun () ->
    expect @@ ZInt.(ediv_rem (of_int (-9)) (of_int (-3))) |> toEqual (ZInt.(of_int (-3), zero )));


  test "is_even" (fun () ->
    expect @@ ZInt.(is_even (of_int (8))) |> toEqual true);  

  test "is_even" (fun () ->
    expect @@ ZInt.(is_even (of_int (9))) |> toEqual false);

  test "is_odd" (fun () ->
    expect @@ ZInt.(is_odd (of_int (8))) |> toEqual false);  

  test "is_odd" (fun () ->
    expect @@ ZInt.(is_odd (of_int (9))) |> toEqual true);
  
  )

File: lib/js/__tests__/bs_Z_Int_test.js
'use strict';

var Z = require("../src/Z.js");
var Jest = require("@glennsl/bs-jest/lib/js/src/jest.js");

describe("Zint", (function () {
        Jest.test("rem", (function () {
                return Jest.Expect[/* toEqual */12](Z.Int[/* of_int */4](1), Jest.Expect[/* expect */0](Z.Int[/* rem */20](Z.Int[/* of_int */4](7), Z.Int[/* of_int */4](3))));
              }));
        Jest.test("rem", (function () {
                return Jest.Expect[/* toEqual */12](Z.Int[/* of_int */4](1), Jest.Expect[/* expect */0](Z.Int[/* rem */20](Z.Int[/* of_int */4](7), Z.Int[/* of_int */4](-3))));
              }));
        Jest.test("rem", (function () {
                return Jest.Expect[/* toEqual */12](Z.Int[/* of_int */4](-1), Jest.Expect[/* expect */0](Z.Int[/* rem */20](Z.Int[/* of_int */4](-7), Z.Int[/* of_int */4](3))));
              }));
        Jest.test("rem", (function () {
                return Jest.Expect[/* toEqual */12](Z.Int[/* of_int */4](-1), Jest.Expect[/* expect */0](Z.Int[/* rem */20](Z.Int[/* of_int */4](-7), Z.Int[/* of_int */4](-3))));
              }));
        Jest.test("ediv_rem 7 3", (function () {
                return Jest.Expect[/* toEqual */12](/* tuple */[
                            Z.Int[/* of_int */4](2),
                            Z.Int[/* of_int */4](1)
                          ], Jest.Expect[/* expect */0](Z.Int[/* ediv_rem */24](Z.Int[/* of_int */4](7), Z.Int[/* of_int */4](3))));
              }));
        Jest.test("ediv_rem 7 (-3)", (function () {
                return Jest.Expect[/* toEqual */12](/* tuple */[
                            Z.Int[/* of_int */4](-2),
                            Z.Int[/* of_int */4](1)
                          ], Jest.Expect[/* expect */0](Z.Int[/* ediv_rem */24](Z.Int[/* of_int */4](7), Z.Int[/* of_int */4](-3))));
              }));
        Jest.test("ediv_rem (-7) 3", (function () {
                return Jest.Expect[/* toEqual */12](/* tuple */[
                            Z.Int[/* of_int */4](-3),
                            Z.Int[/* of_int */4](2)
                          ], Jest.Expect[/* expect */0](Z.Int[/* ediv_rem */24](Z.Int[/* of_int */4](-7), Z.Int[/* of_int */4](3))));
              }));
        Jest.test("ediv_rem (-7) (-3)", (function () {
                return Jest.Expect[/* toEqual */12](/* tuple */[
                            Z.Int[/* of_int */4](-3),
                            Z.Int[/* of_int */4](2)
                          ], Jest.Expect[/* expect */0](Z.Int[/* ediv_rem */24](Z.Int[/* of_int */4](-7), Z.Int[/* of_int */4](-3))));
              }));
        Jest.test("ediv_rem (-8) 3", (function () {
                return Jest.Expect[/* toEqual */12](/* tuple */[
                            Z.Int[/* of_int */4](-3),
                            Z.Int[/* of_int */4](1)
                          ], Jest.Expect[/* expect */0](Z.Int[/* ediv_rem */24](Z.Int[/* of_int */4](-8), Z.Int[/* of_int */4](3))));
              }));
        Jest.test("ediv_rem (-8) (-3)", (function () {
                return Jest.Expect[/* toEqual */12](/* tuple */[
                            Z.Int[/* of_int */4](-3),
                            Z.Int[/* of_int */4](1)
                          ], Jest.Expect[/* expect */0](Z.Int[/* ediv_rem */24](Z.Int[/* of_int */4](-8), Z.Int[/* of_int */4](-3))));
              }));
        Jest.test("ediv_rem 9 3", (function () {
                return Jest.Expect[/* toEqual */12](/* tuple */[
                            Z.Int[/* of_int */4](3),
                            Z.Int[/* of_int */4](0)
                          ], Jest.Expect[/* expect */0](Z.Int[/* ediv_rem */24](Z.Int[/* of_int */4](9), Z.Int[/* of_int */4](3))));
              }));
        Jest.test("ediv_rem 9 (-3)", (function () {
                return Jest.Expect[/* toEqual */12](/* tuple */[
                            Z.Int[/* of_int */4](-3),
                            Z.Int[/* zero */1]
                          ], Jest.Expect[/* expect */0](Z.Int[/* ediv_rem */24](Z.Int[/* of_int */4](9), Z.Int[/* of_int */4](-3))));
              }));
        Jest.test("ediv_rem (-9) 3", (function () {
                return Jest.Expect[/* toEqual */12](/* tuple */[
                            Z.Int[/* of_int */4](-3),
                            Z.Int[/* of_int */4](0)
                          ], Jest.Expect[/* expect */0](Z.Int[/* ediv_rem */24](Z.Int[/* of_int */4](-9), Z.Int[/* of_int */4](3))));
              }));
        Jest.test("ediv_rem (-9) (-3)", (function () {
                return Jest.Expect[/* toEqual */12](/* tuple */[
                            Z.Int[/* of_int */4](-3),
                            Z.Int[/* zero */1]
                          ], Jest.Expect[/* expect */0](Z.Int[/* ediv_rem */24](Z.Int[/* of_int */4](-9), Z.Int[/* of_int */4](-3))));
              }));
        Jest.test("is_even", (function () {
                return Jest.Expect[/* toEqual */12](true, Jest.Expect[/* expect */0](Z.Int[/* is_even */52](Z.Int[/* of_int */4](8))));
              }));
        Jest.test("is_even", (function () {
                return Jest.Expect[/* toEqual */12](false, Jest.Expect[/* expect */0](Z.Int[/* is_even */52](Z.Int[/* of_int */4](9))));
              }));
        Jest.test("is_odd", (function () {
                return Jest.Expect[/* toEqual */12](false, Jest.Expect[/* expect */0](Z.Int[/* is_odd */53](Z.Int[/* of_int */4](8))));
              }));
        Jest.test("is_odd", (function () {
                return Jest.Expect[/* toEqual */12](true, Jest.Expect[/* expect */0](Z.Int[/* is_odd */53](Z.Int[/* of_int */4](9))));
              }));
        Jest.test("gcd 2 12", (function () {
                return Jest.Expect[/* toEqual */12](Z.Int[/* of_int */4](2), Jest.Expect[/* expect */0](Z.Int[/* gcd */54](Z.Int[/* of_int */4](2), Z.Int[/* of_int */4](12))));
              }));
        Jest.test("gcd 12 20", (function () {
                return Jest.Expect[/* toEqual */12](Z.Int[/* of_int */4](4), Jest.Expect[/* expect */0](Z.Int[/* gcd */54](Z.Int[/* of_int */4](12), Z.Int[/* of_int */4](20))));
              }));
        return Jest.test("gcd 36 60", (function () {
                      return Jest.Expect[/* toEqual */12](Z.Int[/* of_int */4](12), Jest.Expect[/* expect */0](Z.Int[/* gcd */54](Z.Int[/* of_int */4](36), Z.Int[/* of_int */4](60))));
                    }));
      }));

/*  Not a pure module */
